File: src/main/java/com/tibiawiki/domain/objects/Effect.java
package com.tibiawiki.domain.objects;

import com.tibiawiki.domain.enums.InfoboxTemplate;
import com.tibiawiki.domain.enums.Status;
import lombok.Builder;
import lombok.Getter;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.List;

@Getter
@Component
public class Effect extends WikiObject {

    private final Integer effectid;
    private final String primarytype;
    private final String secondarytype;
    private final Integer lightradius;
    private final Integer lightcolor;
    private final String causes;
    @SuppressWarnings("squid:S1700") // class and field name are the same, but that's understandable
    private final String effect;

    private Effect() {
        this.effectid = null;
        this.primarytype = null;
        this.secondarytype = null;
        this.lightradius = null;
        this.lightcolor = null;
        this.causes = null;
        this.effect = null;
    }

    @SuppressWarnings("squid:S00107")
    @Builder
    private Effect(String name, String implemented, String notes, String history, Status status, Integer effectid,
                   String primarytype, String secondarytype, Integer lightradius, Integer lightcolor, String causes,
                   String effect) {
        super(name, null, null, null, implemented, notes, history, status);
        this.effectid = effectid;
        this.primarytype = primarytype;
        this.secondarytype = secondarytype;
        this.lightradius = lightradius;
        this.lightcolor = lightcolor;
        this.causes = causes;
        this.effect = effect;
    }

    @Override
    public String getTemplateType() {
        return InfoboxTemplate.EFFECT.getTemplateName();
    }

    @Override
    public List<String> fieldOrder() {
        return Arrays.asList("name", "implemented", "effectid", "primarytype", "secondarytype", "lightradius", "lightcolor",
                "causes", "effect", "notes", "history", "status");
    }
}

File: src/main/kotlin/com/tibiawiki/domain/objects/Effect.kt
package com.tibiawiki.domain.objects

import com.tibiawiki.domain.enums.InfoboxTemplate
import com.tibiawiki.domain.enums.Status

data class Effect(
    override val name: String? = null,
    override val implemented: String? = null,
    override val notes: String? = null,
    override val history: String? = null,
    override val status: Status? = null,
    val effectid: List<Int>? = null,
    val primarytype: String? = null,
    val secondarytype: String? = null,
    val lightradius: Int? = null,
    val lightcolor: Int? = null,
    val causes: String? = null,
    val effect: String? = null
) : WikiObject(
    name = name,
    implemented = implemented,
    notes = notes,
    history = history,
    status = status
) {
    override fun getTemplateType(): String {
        return InfoboxTemplate.EFFECT.templateName
    }

    override fun fieldOrder(): List<String> {
        return listOf(
            "name", "implemented", "effectid", "primarytype", "secondarytype", "lightradius", "lightcolor",
            "causes", "effect", "notes", "history", "status"
        )
    }
}
